In [2]:
import pandas as pd

Загрузка данных и название столбцов (в adult.names)

In [3]:
column_names = ['age', 'workclass', 'fnlwgt', 'education', 'education-num', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'capital-gain',
'capital-loss', 'hours-per-week', 'native-country', 'salary']

data_frame = pd.read_csv('data/adult.data', header=None, names=column_names, na_values='?', skipinitialspace=True)

data_frame.head()


age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

Ответы на вопросы:

1. Число столбцов в наборе данных

In [4]:
print(len(data_frame.columns))

15


2. Есть ли пропуски в данных? Если есть, то в каких столбцах

In [5]:
miss_data = data_frame.isnull().sum()

if miss_data.sum()>0:
    print('Пропуски в следующий столбцах:')
    for col_name in miss_data.index:
        if miss_data[col_name] > 0:
            print('Название столбца: ', col_name, '-- Пропусков: ', miss_data[col_name])
else:
    print("Нет пропусков")

Пропуски в следующий столбцах:
Название столбца:  workclass -- Пропусков:  1836
Название столбца:  occupation -- Пропусков:  1843
Название столбца:  native-country -- Пропусков:  583


3. Кол-во уникальных значений в столбце race

In [6]:
print(data_frame['race'].nunique())
print(data_frame['race'].unique())

5
['White' 'Black' 'Asian-Pac-Islander' 'Amer-Indian-Eskimo' 'Other']


4. Медиана hours-per-week

In [7]:
print(data_frame['hours-per-week'].median())

40.0


5. Кого больше - женщин или мужчин с ЗП >50K?

In [8]:
high_salary = data_frame[data_frame['salary'] == '>50K']

count_sex = high_salary['sex'].value_counts()
print('Мужчин' if count_sex['Male'] > count_sex['Female'] else 'Женщин')


Мужчин


6. Заполните пропущенные данные в отдельных столбцах наиболее встречаемыми значениями. Как еще можно было бы заполнить пропущенные данные?

In [9]:
data_frame_copy = data_frame.copy()

for col_name in miss_data.index:
    if miss_data[col_name] > 0:
        most_use = data_frame_copy[col_name].mode()[0]
        data_frame_copy[col_name] = data_frame_copy[col_name].fillna(most_use)


Проверим что пропусков нет:

In [10]:
miss_data_check = data_frame_copy.isnull().sum()

if miss_data_check.sum()>0:
    print('Пропуски в следующий столбцах:')
    for col_name in miss_data_check.index:
        if miss_data_check[col_name] > 0:
            print('Название столбца: ', col_name, '-- Пропусков: ', miss_data_check[col_name])
else:
    print("Нет пропусков")

Нет пропусков


ОТВЕТЫ:

1. 15
2. Пропуски в следующий столбцах:
Название столбца:  workclass -- Пропусков:  1836
Название столбца:  occupation -- Пропусков:  1843
Название столбца:  native-country -- Пропусков:  583
3. 5
['White' 'Black' 'Asian-Pac-Islander' 'Amer-Indian-Eskimo' 'Other']
4. 40.0
5. Мужчин
6. В зависимости от типа данных. Например пропуски в workclass можно было заполнить "Назначить" или "Проверить" т.к. более используюмую запись не совсем корректно присваивать ко всем пропускам. Occupation - так же как с workclass. Native-country - так же. Это важно, если например нужно понять из какой страны больше зарабатывают, или какое занятие приносит больше дохода. Неверно указанные данные искажают картину, так что лучше эти пропуски заполнять так, чтобы они не учитывались при подобных вычислениях